### PR TITLE
Logging: Defensively access provider resource

### LINKED ISSUE #810: error with logging instrumentation - AttributeError: 'ProxyTracerProvider' object has no attribute 'resource'
**Describe your environment** 

LoggingInstrumentor().instrument() is throwing an error
```
Traceback (most recent call last):
  File "manage.py", line 30, in <module>
    main()
  File "manage.py", line 14, in main
    LoggingInstrumentor().instrument(set_logging_format=True)
  File "/home/vamsikrishnam/otel/lib/python3.8/site-packages/opentelemetry/instrumentation/instrumentor.py", line 109, in instrument
    result = self._instrument(  # pylint: disable=assignment-from-no-return
  File "/home/vamsikrishnam/otel/lib/python3.8/site-packages/opentelemetry/instrumentation/logging/__init__.py", line 81, in _instrument
    resource = provider.resource if provider else None
AttributeError: 'ProxyTracerProvider' object has no attribute 'resource'
```

**Steps to reproduce**
Below packages installed and trying to instrument with below two lines:

>  LoggingInstrumentor().instrument(set_logging_format=True)
>  DjangoInstrumentor().instrument()

```
(otel) vamsikrishnam@NHHYDL-00217:~/django$ pip list | grep opentele
opentelemetry-api                      1.7.1
opentelemetry-exporter-otlp            1.7.1
opentelemetry-exporter-otlp-proto-grpc 1.7.1
opentelemetry-exporter-otlp-proto-http 1.7.1
opentelemetry-instrumentation          0.26b1
opentelemetry-instrumentation-django   0.26b1
opentelemetry-instrumentation-logging  0.26b1
opentelemetry-instrumentation-wsgi     0.26b1
opentelemetry-propagator-b3            1.7.1
opentelemetry-proto                    1.7.1
opentelemetry-sdk                      1.7.1
opentelemetry-semantic-conventions     0.26b1
opentelemetry-util-http                0.26b1
```

**What is the expected behavior?**
What did you expect to see?
logging should be instrumented properly.

**What is the actual behavior?**
What did you see instead?
logging should be instrumented properly and populate the otelTraceID and otelSpanID in the logs.

**Additional context**

$ python3 --version
Python 3.8.10

manage.py:

```
#!/usr/bin/env python
"""Django's command-line utility for administrative tasks."""
import os
import sys
import logging
from opentelemetry.instrumentation.django import DjangoInstrumentor
from opentelemetry.instrumentation.logging import LoggingInstrumentor


def main():
    """Run administrative tasks."""
    os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'django_project.settings')
    logging.basicConfig(level = logging.DEBUG)
    LoggingInstrumentor().instrument(set_logging_format=True)
    DjangoInstrumentor().instrument()
    # LoggingInstrumentor().instrument(set_logging_format=True,log_level=logging.DEBUG)

    try:
        from django.core.management import execute_from_command_line
    except ImportError as exc:
        raise ImportError(
            "Couldn't import Django. Are you sure it's installed and "
            "available on your PYTHONPATH environment variable? Did you "
            "forget to activate a virtual environment?"
        ) from exc
    execute_from_command_line(sys.argv)


if __name__ == '__main__':
    main()
```

### PR DESCRIPTION
**Pull request by owais**
# Description

Now service name is extracted from the provider defensively and lazily.
This accounts for an SDK that does not provide access to "resource" via
TracerProviders and for lazy initialization of TracerProviders.

Fixes #810

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

- [x] Added new test to cover use case
- [x] Manually

# Does This PR Require a Core Repo Change?

- [ ] Yes. - Link to PR: 
- [x] No.

# Checklist:

See [contributing.md](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/CONTRIBUTING.md) for styleguide, changelog guidelines, and more.

- [ ] Followed the style guidelines of this project
- [x] Changelogs have been updated
- [ ] Unit tests have been added
- [ ] Documentation has been updated

### CODE REVIEW

**Review by lzchen — APPROVED** [review]
What does it mean for something to be defensive?

**Comment by owais** [comment]
I guess gracefully downgrade instead of crashing when some condition is not met like using `getattr(.., .., None)` instead of directly accessing an attribute or checking for the presence of something before using it.

File: instrumentation/opentelemetry-instrumentation-logging/src/opentelemetry/instrumentation/logging/__init__.py [hunk 1]
```diff
@@ -76,20 +76,29 @@ def instrumentation_dependencies(self) -> Collection[str]:
...
         old_factory = logging.getLogRecordFactory()
         LoggingInstrumentor._old_factory = old_factory
 
+        service_name = None
+
         def record_factory(*args, **kwargs):
             record = old_factory(*args, **kwargs)
 
             record.otelSpanID = "0"
             record.otelTraceID = "0"
```
Comment (lonewolf3739) [inline instrumentation/opentelemetry-instrumentation-logging/src/opentelemetry/instrumentation/logging/__init__.py:90]: Probably out of scope, what do you think about setting service name to "" here similar to span, traceid and defensively access resource on the below span instead?

Comment (owais) [reply instrumentation/opentelemetry-instrumentation-logging/src/opentelemetry/instrumentation/logging/__init__.py:90]: Not sure about accessing resource from span as span API does not define resource at all. ReadableSpan defines resource which should only be used by the SDK. I know somewhat similar argument applies to provider but for some reason I feel doing it with provider is a lesser evil. Happy to hear other ideas/counter-arguments but may be we shold open an issue and discuss there?

Comment (owais) [reply instrumentation/opentelemetry-instrumentation-logging/src/opentelemetry/instrumentation/logging/__init__.py:90]: Actually we should discuss it in the next SIG call. Not span vs provider but the fact that an instrumentation depends on an SDK feature which shouldn't be the case.

Comment (lonewolf3739) [reply instrumentation/opentelemetry-instrumentation-logging/src/opentelemetry/instrumentation/logging/__init__.py:90]: > the fact that an instrumentation depends on an SDK feature which shouldn't be the case.

Isn't this an exception and I remember there was some discussion around if `instrumentation-logging` should actually be an instrumentation. I am not aware of any other instrumentation depending sdk feature.

**Pull request merged**